crypto/fipsmodule/evp/evp.c

```diff
@@ -445,25 +474,12 @@ int EVP_PKEY_assign(EVP_PKEY *pkey, int type, void *key) {
 
 int EVP_PKEY_set_type(EVP_PKEY *pkey, int type) {
   SET_DIT_AUTO_RESET;
-  if (pkey && pkey->pkey.ptr) {
-    // This isn't strictly necessary, but historically |EVP_PKEY_set_type| would
-    // clear |pkey| even if |evp_pkey_asn1_find| failed, so we preserve that
-    // behavior.
-    free_it(pkey);
-  }
-
-  const EVP_PKEY_ASN1_METHOD *ameth = evp_pkey_asn1_find(type);
-  if (ameth == NULL) {
-    OPENSSL_PUT_ERROR(EVP, EVP_R_UNSUPPORTED_ALGORITHM);
-    ERR_add_error_dataf("algorithm %d", type);
-    return 0;
-  }
-
-  if (pkey) {
-    evp_pkey_set_method(pkey, ameth);
-  }
+  return pkey_set_type(pkey, type, NULL, -1);
+}
 
-  return 1;
+int EVP_PKEY_set_type_str(EVP_PKEY *pkey, const char *str, int len) {
+  SET_DIT_AUTO_RESET;
```

Comment:
@nebeid -- what are the implications of this? is it an appropriate usage?

(i'm not super familiar with DIT directives and their appropriate usage)

Reply:
Yes, it looks good to me, it matches the original behaviour.